Add contribution guidelines (#608)
KhronosGroup/SPIRV-LLVM-Translator: CONTRIBUTING.md

# Contribution guidelines

## If you found a bug or would like to see a new feature

Please reach us by creating a [new issue].

Bug report should include proper description and steps to reproduce:
- attach LLVM BC or SPV file you are trying to translate and command you launch
- any backtrace in case of crashes would be helpful
- please describe what goes wrong or in unexpected way during translation

For feature requests, please describe the feature you would like to see
implemented in the translator.

[new issue]: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/new

## If you would like to contribute your change

Please open a [pull request]. If you are not sure whether your changes are
correct, you can either mark it as [draft] or create an issue to discuss the

> problem and possible ways to fix it prior publishing a PR.

Reviewer: ```suggestion
problem and possible ways to fix it prior to publishing a PR.
```

It is okay to have several commits in the PR, but each of them should be

> build-able and tests should pass: repo maintainers can squash several commits

Reviewer: ```suggestion
buildable and tests should pass. Maintainers can squash several commits
```

into a single one for you during merge, but if you would like to see several
commits in the git history, please let us know in PR description/comments so
maintainers will rebase your PR instead of squashing it.

Each functional change (new feature or bug fix) must be supplied with
corresponding tests. See [#testing-guidelines] for more information about
testing. NFC (non-functional change) PRs can be accepted without new tests.

Code changes should follow coding standards, which are inherited from [LLVM

> Coding Standards] - compliance of your code is checked automatically using

Reviewer: ```suggestion
Coding Standards]. Compliance of your code is checked automatically using
```

Travis CI. See [clang-format] and [clang-tidy] configs for more details about
coding standards.

In order to get your PR merged, the following conditions must be met:

> - If you are first-time contributor, you have to sign

Reviewer: ```suggestion
- If you are a first-time contributor, you have to sign the
```

  [Contributor License Agreement]. Corresponding link and instructions will be
  automatically posted into your PR.
- [Travis CI testing] jobs must pass on your PR: this includes functional
  testing and checking for complying with coding standards.

> - You need to get approval from at least one contributor with merge rights

Reviewer: ```suggestion
- You need to get approval from at least one contributor with merge rights.
```

> As a contributor, you should expect that even approved PR might still be left

Reviewer: ```suggestion
As a contributor, you should expect that even an approved PR might still be left
```

unmerged for a few days: this is needed, because the translator is being
developed by different vendors and individuals and we need to ensure that each

> interested party were able to react to new changes and provide feedback.

Reviewer: ```suggestion
interested party is able to react to new changes and provide feedback.
```

Information below is a guideline for repo maintainers and can be used by

> contributors to get some expectations about how long PR has to be opened before

Reviewer: ```suggestion
contributors to get some expectations about how long a PR has to be open before
```

it can be merged:

> - For any significant change/redesign, the PR must be opened for at least 5

Reviewer: ```suggestion
- For any significant change/redesign, the PR must be open for at least 5
```

  working days, so everyone interested can step in to provide feedback, discuss

>   direction and help to find bugs

Reviewer: ```suggestion
  direction and help to find bugs.
```

>   - Ideally, there should be more several approvals from different

Reviewer: ```suggestion
  - Ideally, there should be approvals from different
```

>     vendors/individuals to get it merged

Reviewer: ```suggestion
    vendors/individuals to get it merged, particularly for larger changes.
```

> - For regular changes/bug fixes, the PR must be opened for at least 2-3 working

Reviewer: ```suggestion
- For regular changes/bug fixes, the PR must be open for at least 2-3 working
```

>   days, so everyone interested can step in for review and providing feedback

Reviewer: ```suggestion
  days, so everyone interested can step in for review and provide feedback.
```

>   - If change is vendor-specific (bug fix in vendor extension implementation or

Reviewer: ```suggestion
  - If the change is vendor-specific (bug fix in vendor extension implementation or
```

    new vendor-specific extension support), then it is okay to merge PR sooner

>   - If change affects or might affect several interesting parties, the PR must

Reviewer: ```suggestion
  - If the change affects or might affect several interested parties, the PR must
```

>     be left opened for 2-3 working days and it would be good to see feedback

Reviewer: ```suggestion
    be left open for 2-3 working days and it would be good to see feedback
```

    from different vendors/inviduals before merging
- Tiny NFC changes or trivial build fixes (due to LLVM API changes) can be
  submitted as soon as testing is finished and PR approved - no need to wait for

>   too long

Reviewer: ```suggestion
  too long.
```

- In general, just use common sense to wait long enough to get feedback from
  everyone who might be interested in the PR and don't hesitate to explicitly

>   mention ones who might be interested in reviewing the PR

Reviewer: ```suggestion
  mention individuals who might be interested in reviewing the PR.
```
